pkgs/applications/version-management/p4/default.nix

```diff
@@ -1,31 +1,108 @@
+, libcxx
+, libcxxabi
+, openssl
+, xcbuild
+, CoreServices
+, Foundation
+, Security
+}:
 
-stdenv.mkDerivation rec {
+let
+  opensslStatic = openssl.override { static = true; };
+  androidZlibContrib =
+    let
+      src = fetchzip {
+        url = "https://android.googlesource.com/platform/external/zlib/+archive/61174f4fd262c6075f88768465f308aae95a2f04.tar.gz";
+        sha256 = "sha256-EMzKAHcEWOUugcHKH2Fj3ZaIHC9UlgO4ULKe3RvgxvI=";
+        stripRoot = false;
+      };
+    in linkFarm "android-zlib-contrib" [
+      # We only want to keep the contrib directory as the other files conflict
+      # with p4's own zlib files. (For the same reason, we can't use the
+      # cone-based Git sparse checkout, either.)
+      { name = "contrib"; path = "${src}/contrib"; }
+    ];
+  libcxxUnified = symlinkJoin {
+    inherit (libcxx) name;
+    paths = [ libcxx libcxxabi ];
+  };
+
+  version = "2022.1.2305383";
+in
+stdenv.mkDerivation {
   pname = "p4";
-  version = "2021.2.2201121";
+  inherit version;
 
   src = fetchurl {
-    # actually  https://cdist2.perforce.com/perforce/r21.2/bin.linux26x86_64/helix-core-server.tgz but upstream deletes releases
-    url = "https://web.archive.org/web/20211118024943/https://cdist2.perforce.com/perforce/r21.2/bin.linux26x86_64/helix-core-server.tgz";
-    sha256 = "sha256-SrfI2ZD7KDyttCd8+fo8g4UZKljYYO/SbzqrS9tAcC8=";
+    # Upstream replaces minor versions, so use cached URL.
+    url = "https://web.archive.org/web/20220901184735/https://ftp.perforce.com/perforce/r22.1/bin.tools/p4source.tgz";
+    sha256 = "27ab3ddd7b178b05cf0b710e941650dac0688d294110ebafda9027732c0944c6";
   };
 
-  sourceRoot = ".";
+  buildInputs = lib.optionals stdenv.isDarwin [ CoreServices Foundation Security ];
+  nativeBuildInputs = [ jam ];
+
+  hardeningDisable = lib.optionals stdenv.isDarwin [ "strictoverflow" ];
 
-  dontBuild = true;
+  # Flags understood by the Jamfile and imported from the environment:
+  CROSS_COMPILE = stdenv.cc.targetPrefix;
+  CCFLAGS =
+    # The file contrib/optimizations/slide_hash_neon.h is missing from the
+    # upstream distribution. It comes from the Android/Chromium sources.
+    lib.optionals stdenv.isAarch64 [ "-I${androidZlibContrib}" ];
+  "C++FLAGS" =
+    # Avoid a compilation error that only occurs for 4-byte longs.
+    lib.optionals stdenv.isi686 [ "-Wno-narrowing" ]
+    # See the "Header dependency changes" section of
+    # https://www.gnu.org/software/gcc/gcc-11/porting_to.html for more information
+    # on why we need to include these.
+    ++ lib.optionals
+      (stdenv.cc.isClang || (stdenv.cc.isGNU && lib.versionAtLeast stdenv.cc.cc.version "11.0.0"))
+      [ "-include" "limits" "-include" "thread" ];
+  MALLOC_OVERRIDE = "no";
+  SSLINCDIR = "${lib.getDev opensslStatic}/include";
+  SSLLIBDIR = "${lib.getLib opensslStatic}/lib";
+  MACOSX_SDK = lib.optionalString stdenv.isDarwin "${emptyDirectory}";
+  "LIBC++DIR" = lib.optionalString stdenv.isDarwin "${libcxxUnified}/lib";
```

Comment:
Those should be exportet in some prePhase like preBuild

Reply:
These are all actually flags to `jam`, so I'm just going to move them into there. The way I have them split out right now doesn't really make any sense I don't think.

Reply:
I moved all of them except CCFLAGS and C++FLAGS which are Jam's standard equivalent to the autotools/make convention of CFLAGS and CXXFLAGS, respectively, so they felt like they could stay at the top level. Since they're built from a list of conditionals it's also harder to put them inside a `preBuild`-style script that maintains readability.